sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,363 @@
+# Admission control changes
+
+Admission control is the key security and cluster policy enforcement mechanism
+in Kubernetes. In order to make a reliable extension mechanism for admission,
+we must reconsider the current chain to produce understandable extensions.
+
+We are proposing a series of changes to admission control in an attempt to make
+it less error-prone, easier to configure, and more flexible for admission plugin
+developers.
+
+## Requirements for extensible admission
+Ordering the admission chain is critical.  One mutating admission plugin can invalidate
+the results of other admission plugins.  For instance, the `LimitVerifier` would
+invalidate results for the `ResourceQuota` plugin.  More generally, object mutations
+must happen before object validations.  In addition, compiled validating admission plugins
+must run after mutating admission webhooks.  Splitting the chain into mutators and validators
+will resolve the coarse grained ordering problem and would require two different webhook APIs
+since the `Status` would be different for both now and we can reasonably expect different
+evolutions.  It would also force changes in the registration APIs since ordering matters
+for individual mutators, but ordering does not matter for validators.  Having fewer items to 
+order makes the registration API easier to reason about for maintainability.
+
+Initializers as implemented today attempt to push new API validation requirements into the 
+strategy and new requirements onto the admission plugins handling updates and creates.  The
+former is unnecessary and the latter doesn't work.
+
+API validation on uninitialized objects can reasonably follow the rules for validation on create
+plus `ValidateObjectMetaUpdate` (to prevent name and UID changes and the like).
+Since the object hasn't been "ready", there is no reason to add the validation rules required on
+updates since restrictions like field immutability don't logically apply and would restrict the power
+of admission extensions to the point of unusability.  Using the results for creation validation on 
+uninitialized objects and objects transitioning from uninitialized to initialized lets the
+strategy author focus on their API needs.
```

Comment:
> PodStatus shouldn't be updated when the pod is not initialized

that's not enforced by validation... that's done in PrepareForUpdate

Reply:
Do you mean disallow any UPDATE to subresources when the object is not initialized? That sounds right.

@deads2k it sounds like you don't want to pursue the "uninitialized objects can be invalid" argument anymore?

Reply:
> @deads2k it sounds like you don't want to pursue the "uninitialized objects can be invalid" argument anymore?

Oh, I'd still like that and I still think it would be correct to allow it, but that decision isn't required to make the API beta

Reply:
@deads2k, agree it's not required for beta. One way to allow "invalid" uninitialized objects is introducing the `ValidateUninitialized` method, so that strategy author can further relax the validation. Does that sound reasonable?

Reply:
> @deads2k, agree it's not required for beta. One way to allow "invalid" uninitialized objects is introducing the ValidateUninitialized method, so that strategy author can further relax the validation. Does that sound reasonable?

As long as the normal flow runs the standard create validation (not update validation), then I'm ok having an opt-in method.  The vast majority of resources won't care about specific initialization logic if run the create validation for them instead of the update validation.